dvc/remote/base.py

```diff
@@ -685,17 +685,21 @@ def _checkout_file(
         self, path_info, checksum, force, progress_callback=None
     ):
         cache_info = self.checksum_to_path_info(checksum)
-        if self.exists(path_info):
-            msg = "data '{}' exists. Removing before checkout."
-            logger.warning(msg.format(str(path_info)))
-            self.safe_remove(path_info, force=force)
 
-        self.link(cache_info, path_info)
-        self.state.save_link(path_info)
-        self.state.save(path_info, checksum)
+        if self._needs_checkout(path_info, checksum, force):
```

Comment:
```
        # Relinking with `symlink/hardlink/reflink` is cheap, but if we need `copy` links, then to save
        # on copying we should only relink if we already have `hardlink` or a `symlink`, we don't care
        # about reflinks, because they are indistinguishable from copy anyway.
        if self._is_cache_type_copy() \
            and not self.changed(path_info, checksum_info) \
            and not System.is_hardlink(path_info) \
            and not System.is_symlink(path_info):
                # protecting/unprotecting is cheap, so there is no reason to detect it
                if self.protected:
                   self.protect(path_info)
                else:
                   self.unprotect(path_info)
            msg = "File '{}' didn't change."                         
            logger.debug(msg.format(str(path_info)))                 
        else:
             if self.exists(path_info):
                 ...
                 self.safe_remove(path_info, force=force)
            self.link(cache_info, path_info)
            ...
```

Reply:
done